packages/core/src/fhir-deduplication/deduplicate-fhir.ts

```diff
@@ -29,350 +30,285 @@ import { createRef } from "./shared";
-  ["Immunization", immunizationFiltersMap],
-  ["MedicationStatement", medicationRelatedFiltersMap],
-  ["MedicationRequest", medicationRelatedFiltersMap],
-  ["MedicationAdministration", medicationRelatedFiltersMap],
-  ["Composition", compositionFiltersMap],
-  ["Observation", observationFiltersMap],
-  ["Procedure", procedureFiltersMap],
-]);
-
 export function removeResourcesWithDanglingLinks(
-  entries: BundleEntry<Resource>[],
-  links: string[]
-) {
-  return entries.flatMap(entry => handleDanglingLinks(entry as Resource, links));
-}
+  resourceArrays: ExtractedFhirTypes,
+  danglingLinks: Set<string>
+): { updatedResourceArrays: ExtractedFhirTypes; deletedRefs: Set<string> } {
+  const deletedRefs = new Set<string>();
+  const updatedResourceArrays = initExtractedFhirTypes();
 
-function handleDanglingLinks(res: Resource, danglingLinks: string[]): Resource | [] {
-  if (res) {
-    let entry = res;
-    const filtersMap = resourceFiltersMap.get(res.resourceType);
-
-    if (filtersMap) {
-      for (const danglingLink of danglingLinks) {
-        const linkResourceType =
-          res.resourceType === "Composition" ? "all" : danglingLink.split("/")[0];
-        if (linkResourceType) {
-          const callbackFn = filtersMap.get(linkResourceType);
-          if (callbackFn) {
-            const result = callbackFn(entry, danglingLink);
-            if (result) entry = result;
-            else {
-              // removed resources become dangling links
-              danglingLinks.push(createRef(entry));
-              return [];
-            }
-          }
+  for (const [key, resources] of Object.entries(resourceArrays)) {
+    if (Array.isArray(resources)) {
+      const updatedResources: Resource[] = [];
+      for (const resource of resources) {
+        const result = removeDanglingReferences(resource, danglingLinks);
+        if (!result) {
+          const ref = createRef(resource);
+          if (ref) deletedRefs.add(ref);
+        } else {
+          updatedResources.push(result);
         }
       }
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      updatedResourceArrays[key as keyof ExtractedFhirTypes] = updatedResources as any;
     }
-    return entry;
   }
-  return [];
-}
 
-function removeResource<T extends Resource>(entry: T, link: string): T | undefined {
-  if (JSON.stringify(entry).includes(link)) return undefined;
-  return entry;
+  return { updatedResourceArrays, deletedRefs };
 }
 
-function removeDanglingReferences<T extends Resource>(entry: T, link: string): T {
+function removeDanglingReferences<T extends Resource>(
+  entry: T,
+  danglingLinks: Set<string>
+): T | undefined {
   if (!entry) return entry;
-  if ("result" in entry) {
-    const results = entry.result;
-    if (Array.isArray(results)) {
-      entry.result = results.filter(entry => entry.reference !== link);
-      if (!entry.result.length) delete entry.result;
+
+  const medicationRelatedTypes = [
```

Comment:
put this outside of the function, so we don't create this on every function call